internal/cmd/tui.go

```diff
@@ -0,0 +1,263 @@
+		m.scrollBy(delta)
+	}
+}
+
+const (
+	tab               = "  "
+	defaultNumEntries = 5
+)
+
+func (m tuiModel) View() string {
+	s := fmt.Sprintf(
+		"%s%s %s",
+		ansi.Color("run", "red+b"),
+		ansi.Color("me", "white+b"),
+		ansi.Color(m.version, "white+d"),
+	)
+
+	s += "\n\n"
+
+	for i := m.scroll; i < m.scroll+m.numBlocksShown(); i++ {
+		block := m.blocks[i]
+
+		active := i == m.cursor
+		_, expanded := m.expanded[i]
+
+		line := " "
+		if active {
+			line = ">"
+		}
+
+		line += " "
+
+		{
+			name := block.Name()
+			lang := ansi.Color(block.Language(), "white+d")
+
+			if active {
+				name = ansi.Color(name, "white+b")
+			} else {
+				lang = ""
+			}
+
+			identifier := fmt.Sprintf(
+				"%s %s",
+				name,
+				lang,
+			)
+
+			line += identifier + "\n"
+		}
+
+		codeLines := block.Lines()
+
+		for i, codeLine := range codeLines {
+			content := tab + tab + codeLine
+
+			if !expanded && len(codeLines) > 1 {
+				content += " (...)"
+			}
+
+			content = ansi.Color(content, "white+d")
+
+			if i >= 1 && !expanded {
+				break
+			}
+
+			line += content + "\n"
+		}
+
+		s += line
+	}
+
+	s += "\n"
+
+	{
+		help := strings.Join(
+			[]string{
+				fmt.Sprintf("%v/%v", m.cursor+1, len(m.blocks)),
+				"Choose ↑↓←→",
+				"Run [Enter]",
+				"Expand [Space]",
+				"Quit [^C]",
+				"  by Stateful",
+			},
+			tab,
+		)
+
+		help = ansi.Color(help, "white+d")
+
+		s += help
+	}
+
+	return s
+}
+
+func (m tuiModel) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
+	keyMsg, isKeyPress := msg.(tea.KeyMsg)
+
+	if isKeyPress {
+		switch keyMsg.String() {
+		case "ctrl+c", "q":
+			m.result = tuiResult{
+				exit: true,
+			}
+
+			return m, tea.Quit
+
+		case "up", "k":
+			m.moveCursor(-1)
+
+		case "down", "j":
+			m.moveCursor(1)
+
+		case " ":
+			if _, ok := m.expanded[m.cursor]; ok {
+				delete(m.expanded, m.cursor)
+			} else {
+				m.expanded[m.cursor] = struct{}{}
+			}
+
+		case "enter", "l":
+			m.result = tuiResult{
+				run: m.blocks[m.cursor],
+			}
+
+			return m, tea.Quit
+		}
+	}
+
+	return m, nil
+}
+
+func tuiCmd() *cobra.Command {
+	numEntries := 5
+	exitAfterRun := false
+
+	cmd := cobra.Command{
+		Use:   "tui",
+		Short: "Run the interactive TUI",
+		Long:  "Run a command from a descriptive list given by an interactive TUI",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			blocks, err := getCodeBlocks()
+			if err != nil {
+				return err
+			}
+
+			version := "???"
+
+			bi, ok := debug.ReadBuildInfo()
+			if ok {
+				version = bi.Main.Version
+			}
+
+			if numEntries <= 0 {
+				numEntries = goMath.MaxInt32
+			}
+
+			model := tuiModel{
+				blocks:     blocks,
+				version:    version,
+				expanded:   make(map[int]struct{}),
+				numEntries: numEntries,
+			}
+
+			for {
+				prog := tea.NewProgram(model)
+
+				newModel, err := prog.Run()
+				if err != nil {
+					return err
+				}
+
+				model = newModel.(tuiModel)
+				result := model.result
+
+				if result.run != nil {
```

Comment:
You can use a technique called the early return to avoid unnecessary nesting:

```go
if result.run != nil {
    break
}

if err := runBlockCmd(result.run, cmd, nil); err != nil {
   ...
}
```